File: app/src/main/java/it/amonshore/comikkua/ui/comics/ComicsDetailFragment.java
package it.amonshore.comikkua.ui.comics;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.transition.TransitionInflater;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.view.ActionMode;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;
import androidx.recyclerview.selection.SelectionTracker;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import it.amonshore.comikkua.Constants;
import it.amonshore.comikkua.DateFormatterHelper;
import it.amonshore.comikkua.LiveDataEx;
import it.amonshore.comikkua.LogHelper;
import it.amonshore.comikkua.R;
import it.amonshore.comikkua.data.comics.ComicsViewModel;
import it.amonshore.comikkua.data.comics.ComicsWithReleases;
import it.amonshore.comikkua.data.release.ComicsRelease;
import it.amonshore.comikkua.data.release.Release;
import it.amonshore.comikkua.data.release.ReleaseViewModel;
import it.amonshore.comikkua.ui.ActionModeController;
import it.amonshore.comikkua.ui.DrawableTextViewTarget;
import it.amonshore.comikkua.ui.ImageHelper;
import it.amonshore.comikkua.ui.OnNavigationFragmentListener;
import it.amonshore.comikkua.ui.ShareHelper;
import it.amonshore.comikkua.ui.releases.ReleaseAdapter;


public class ComicsDetailFragment extends Fragment {

    private OnNavigationFragmentListener mListener;

    private TextView mInitial, mName, mPublisher, mAuthors, mNotes,
            mLast, mNext, mMissing;
    private ReleaseAdapter mAdapter;
    private RecyclerView mRecyclerView;
    private SwipeRefreshLayout mSwipeRefreshLayout;

    private ComicsViewModel mComicsViewModel;
    private ReleaseViewModel mReleaseViewModel;
    private long mComicsId;
    private ComicsWithReleases mComics;

    public ComicsDetailFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setSharedElementEnterTransition(TransitionInflater.from(getContext())
                .inflateTransition(android.R.transition.move));
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_comics_detail, container, false);

        final String actionModeName = getClass().getSimpleName() + "_actionMode";
        final Context context = requireContext();
        mRecyclerView = view.findViewById(R.id.list);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(context));

        mSwipeRefreshLayout = view.findViewById(R.id.swipe_refresh);
        mSwipeRefreshLayout.setOnRefreshListener(this::performUpdate);

        mComicsId = ComicsDetailFragmentArgs.fromBundle(requireArguments()).getComicsId();
        // lo stesso nome della transizione è stato assegnato alla view di partenza
        //  il nome deve essere univoco altrimenti il meccanismo non saprebbe quali viste animare
        view.findViewById(R.id.comics).setTransitionName("comics_tx_" + mComicsId);

        mInitial = view.findViewById(R.id.txt_comics_initial);
        mName = view.findViewById(R.id.txt_comics_name);
        mPublisher = view.findViewById(R.id.txt_comics_publisher);
        mAuthors = view.findViewById(R.id.txt_comics_authors);
        mNotes = view.findViewById(R.id.txt_comics_notes);
        mLast = view.findViewById(R.id.txt_comics_release_last);
        mNext = view.findViewById(R.id.txt_comics_release_next);
        mMissing = view.findViewById(R.id.txt_comics_release_missing);

        final ActionModeController actionModeController = new ActionModeController(R.menu.menu_releases_selected) {
            @Override
            public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
                final SelectionTracker<Long> tracker = mAdapter.getSelectionTracker();
                switch (item.getItemId()) {
                    case R.id.purchaseReleases:
                        if (tracker.hasSelection()) {
                            // TODO: considerare le multi release
                            mReleaseViewModel.togglePurchased(tracker.getSelection());
                        }
                        // mantengo la selezione
                        return true;
                    case R.id.orderReleases:
                        if (tracker.hasSelection()) {
                            // TODO: considerare le multi release
                            mReleaseViewModel.toggleOrdered(tracker.getSelection());
                        }
                        // mantengo la selezione
                        return true;
                    case R.id.deleteReleases:
                        if (tracker.hasSelection()) {
                            // prima elimino eventuali release ancora in fase di undo
                            mReleaseViewModel.deleteRemoved();
                            mReleaseViewModel.remove(tracker.getSelection(), count -> showUndo(count));
                        }
                        tracker.clearSelection();
                        return true;
                    case R.id.shareReleases:
                        LiveDataEx.observeOnce(mReleaseViewModel.getComicsReleases(tracker.getSelection()), getViewLifecycleOwner(),
                                items -> ShareHelper.shareReleases(requireActivity(), items));
                        // mantengo la selezione
                        return true;
                }
                return false;
            }

            @Override
            public void onDestroyActionMode(ActionMode mode) {
                // action mode distrutta (anche con BACK, che viene gestito internamente all'ActionMode e non può essere evitato)
                mAdapter.getSelectionTracker().clearSelection();
                super.onDestroyActionMode(mode);
            }
        };

        mAdapter = new ReleaseAdapter.Builder(mRecyclerView)
                .withOnItemSelectedListener((keys, size) -> {
                    if (mListener != null) {
                        if (size == 0) {
                            mListener.onFragmentRequestActionMode(null, actionModeName, null);
                        } else {
                            mListener.onFragmentRequestActionMode(actionModeController, actionModeName,
                                    getString(R.string.title_selected, size));
                        }
                    }
                })
                .withReleaseCallback(new ReleaseAdapter.ReleaseCallback() {
                    @Override
                    public void onReleaseClick(@NonNull ComicsRelease release) {
                        openEdit(view, release);
                    }

                    @Override
                    public void onReleaseTogglePurchase(@NonNull ComicsRelease release) {
                        mReleaseViewModel.updatePurchased(!release.release.purchased, release.release.id);
                    }

                    @Override
                    public void onReleaseToggleOrder(@NonNull ComicsRelease release) {
                        mReleaseViewModel.updateOrdered(!release.release.ordered, release.release.id);
                    }

                    @Override
                    public void onReleaseMenuSelected(@NonNull ComicsRelease release) {
                        // non gestito
                    }
                })
                // uso la versione "lite" con il layout per gli item più compatta
                .useLite()
                .build();

        // recupero il ViewModel per l'accesso ai dati
        mComicsViewModel = new ViewModelProvider(requireActivity())
                .get(ComicsViewModel.class);
        // TODO: lo devo tenere legato al fragment e non all'activity?
        mReleaseViewModel = new ViewModelProvider(this)
                .get(ReleaseViewModel.class);
        // recupero l'istanza e poi rimuovo subito l'observer altrimenti verrebbe notificato il salvataggio
        // TODO: beh non dovrei cmq aggiornare i tre contatori?
        mComicsViewModel.getComicsWithReleases(mComicsId).observe(getViewLifecycleOwner(), comics -> {
            mComics = comics;
            if (comics != null) {
//                mInitial.setText(comics.comics.getInitial());
                mName.setText(comics.comics.name);
                mPublisher.setText(comics.comics.publisher);
                mAuthors.setText(comics.comics.authors);
                mNotes.setText(comics.comics.notes);

                if (comics.comics.hasImage()) {
                    mInitial.setText("");
                    Glide.with(this)
                            .load(Uri.parse(comics.comics.image))
                            .apply(ImageHelper.getGlideCircleOptions())
                            .into(new DrawableTextViewTarget(mInitial));
                } else {
                    mInitial.setText(comics.comics.getInitial());
                }

                final Release lastRelease = comics.getLastPurchasedRelease();
                mLast.setText(lastRelease == null ? context.getString(R.string.release_last_none) :
                        context.getString(R.string.release_last, lastRelease.number));

                final Release nextRelease = comics.getNextToPurchaseRelease();
                if (nextRelease != null) {
                    if (nextRelease.date != null) {
                        // TODO: non mi piace, dovrei mostrare la data solo se futura e nel formato ddd dd MMM
                        mNext.setText(context.getString(R.string.release_next_dated, nextRelease.number,
                                DateFormatterHelper.toHumanReadable(context, nextRelease.date, DateFormatterHelper.STYLE_SHORT)));
                    } else {
                        mNext.setText(context.getString(R.string.release_next, nextRelease.number));
                    }
                } else {
                    mNext.setText(context.getString(R.string.release_next_none));
                }

                final int missingCount = comics.getNotPurchasedReleaseCount();
                mMissing.setText(context.getString(R.string.release_missing, missingCount));
            }
        });

        mReleaseViewModel.getReleaseViewModelItems(mComicsId).observe(getViewLifecycleOwner(), items -> {
            LogHelper.d("release viewmodel data changed size:" + items.size());
            mAdapter.submitList(items);
        });

        // ripristino la selezione salvata in onSaveInstanceState
        mAdapter.getSelectionTracker().onRestoreInstanceState(savedInstanceState);

        return view;
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (context instanceof OnNavigationFragmentListener) {
            mListener = (OnNavigationFragmentListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnNavigationFragmentListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (mAdapter != null) {
            // ripristino le selezioni
            mAdapter.getSelectionTracker().onSaveInstanceState(outState);
        }
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.menu_comics_detail, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.editComics:
                Navigation.findNavController(requireView())
                        .navigate(ComicsDetailFragmentDirections
                                .actionDestComicsDetailFragmentToComicsEditFragment()
                                .setComicsId(mComicsId));

                return true;
            case R.id.createNewRelease:
                Navigation.findNavController(requireView())
                        .navigate(ComicsDetailFragmentDirections
                                .actionDestComicsDetailFragmentToReleaseEditFragment(mComicsId)
                                .setSubtitle(R.string.title_release_create));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void performUpdate() {
        // annullo eventuale undo
        mListener.dismissSnackbar();
        mSwipeRefreshLayout.setRefreshing(true);

        // cerco tutte le nuove release e le aggiungo direttamente
        mReleaseViewModel.getNewReleases(mComics).observe(getViewLifecycleOwner(), releases -> {
            if (releases != null) {
                final int size = releases.size();
                if (size > 0) {
                    mReleaseViewModel.insert(releases.toArray(new Release[size]));
                    Toast.makeText(requireContext(),
                            getResources().getQuantityString(R.plurals.auto_update_available_message, size, size),
                            Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(requireContext(), R.string.auto_update_zero, Toast.LENGTH_SHORT).show();
                }
            } else {
                // in realtà è null in caso di errore
                Toast.makeText(requireContext(), R.string.auto_update_zero, Toast.LENGTH_SHORT).show();
            }
            mSwipeRefreshLayout.setRefreshing(false);
        });
    }

    private void openEdit(@NonNull View view, @NonNull ComicsRelease release) {
        final NavDirections directions = ComicsDetailFragmentDirections
                .actionDestComicsDetailFragmentToReleaseEditFragment(release.comics.id)
                .setReleaseId(release.release.id);

        Navigation.findNavController(view).navigate(directions);
    }

    private void showUndo(int count) {
        mListener.requestSnackbar(getResources().getQuantityString(R.plurals.release_deleted, count, count),
                Constants.UNDO_TIMEOUT,
                (canDelete) -> {
                    if (canDelete) {
                        LogHelper.d("Delete removed releases");
                        mReleaseViewModel.deleteRemoved();
                    } else {
                        LogHelper.d("Undo removed releases");
                        mReleaseViewModel.undoRemoved();
                    }
                });
    }
}

File: app/src/main/java/it/amonshore/comikkua/ui/comics/ComicsDetailFragment.kt
package it.amonshore.comikkua.ui.comics

import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.view.ActionMode
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.Observer
import androidx.navigation.NavDirections
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import it.amonshore.comikkua.BuildConfig
import it.amonshore.comikkua.LogHelper
import it.amonshore.comikkua.R
import it.amonshore.comikkua.data.comics.ComicsWithReleases
import it.amonshore.comikkua.data.release.ComicsRelease
import it.amonshore.comikkua.databinding.FragmentComicsDetailBinding
import it.amonshore.comikkua.toHumanReadable
import it.amonshore.comikkua.ui.DrawableTextViewTarget
import it.amonshore.comikkua.ui.ImageHelperKt
import it.amonshore.comikkua.ui.OnNavigationFragmentListener
import it.amonshore.comikkua.ui.createActionModeCallback
import it.amonshore.comikkua.ui.releases.adapter.ReleaseAdapter
import it.amonshore.comikkua.ui.share

private val ACTION_MODE_NAME = ComicsDetailFragment::class.java.simpleName + "_actionMode"

class ComicsDetailFragment : Fragment() {

    private val _viewModel: ComicsDetailViewModel by viewModels()

    private val _comicsId: Long by lazy {
        ComicsDetailFragmentArgs.fromBundle(requireArguments()).comicsId
    }

    private lateinit var _listener: OnNavigationFragmentListener
    private lateinit var _adapter: ReleaseAdapter
    private lateinit var _comics: ComicsWithReleases

    private var _binding: FragmentComicsDetailBinding? = null
    private val binding get() = _binding!!

//    override fun onCreate(savedInstanceState: Bundle?) {
//        super.onCreate(savedInstanceState)
//        sharedElementEnterTransition = TransitionInflater.from(context)
//            .inflateTransition(android.R.transition.move)
//    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentComicsDetailBinding.inflate(layoutInflater, container, false)
        binding.list.layoutManager = LinearLayoutManager(requireContext())
        binding.swipeRefresh.setOnRefreshListener(::loadNewReleases)

//        // lo stesso nome della transizione è stato assegnato alla view di partenza
//        //  il nome deve essere univoco altrimenti il meccanismo non saprebbe quali viste animare
//        view.findViewById<View>(R.id.comics).transitionName = "comics_tx_$_comicsId"

        val actionModeCallback = createActionModeCallback()
        _adapter = createReleaseAdapter(actionModeCallback)

        _viewModel.getComicsWithReleases(_comicsId)
            .observe(viewLifecycleOwner, createComicsWithReleasesObserver())

        _viewModel.getReleaseViewModelItems(_comicsId)
            .observe(viewLifecycleOwner) { items ->
                LogHelper.d { "release view model data changed size=${items.size}" }
                _adapter.submitList(items)
            }

        _viewModel.events.observe(viewLifecycleOwner) { result ->
            when (result) {
                is UiComicsDetailEvent.MarkedAsRemoved -> onMarkedAsRemoved(
                    result.count,
                    result.tag
                )

                is UiComicsDetailEvent.Sharing -> shareReleases(result.releases)
                is UiComicsDetailEvent.NewReleasesLoaded -> onNewReleasesLoaded(
                    result.count,
                    result.tag
                )

                is UiComicsDetailEvent.NewReleasesError -> onNewReleasesError()
            }
        }

        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupMenu()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        _listener = if (context is OnNavigationFragmentListener) {
            context
        } else {
            throw RuntimeException("$context must implement OnNavigationFragmentListener")
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        // ripristino le selezioni
        _adapter.selectionTracker.onSaveInstanceState(outState)
    }

    private fun setupMenu() {
        val menuHost: MenuHost = requireActivity()
        menuHost.addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.menu_comics_detail, menu)
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                return when (menuItem.itemId) {
                    R.id.editComics -> {
                        findNavController(requireView())
                            .navigate(
                                ComicsDetailFragmentDirections
                                    .actionDestComicsDetailFragmentToComicsEditFragment()
                                    .setComicsId(_comicsId)
                            )
                        true
                    }

                    R.id.createNewRelease -> {
                        findNavController(requireView())
                            .navigate(
                                ComicsDetailFragmentDirections
                                    .actionDestComicsDetailFragmentToReleaseEditFragment(_comicsId)
                                    .setSubtitle(R.string.title_release_create)
                            )
                        true
                    }

                    else -> false
                }
            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)
    }

    private fun createActionModeCallback(): ActionMode.Callback {
        return createActionModeCallback(
            menuRes = R.menu.menu_releases_selected,
            onAction = { actionId: Int ->
                with(_adapter.selectionTracker) {
                    when (actionId) {
                        R.id.purchaseReleases -> {
                            // TODO: considerare le multi release
                            _viewModel.togglePurchased(selection.toList())
                            true
                        }

                        R.id.orderReleases -> {
                            // TODO: considerare le multi release
                            _viewModel.toggleOrdered(selection.toList())
                            true
                        }

                        R.id.deleteReleases -> {
                            _viewModel.markAsRemoved(selection.toList())
                            clearSelection()
                            true
                        }

                        R.id.shareReleases -> {
                            _viewModel.getShareableComicsReleases(selection.toList())
                            true
                        }

                        else -> false
                    }
                }
            },
            onDestroy = {
                // action mode distrutta (anche con BACK, che viene gestito internamente all'ActionMode e non può essere evitato)
                _adapter.selectionTracker.clearSelection()
            })
    }

    private fun createReleaseAdapter(
        actionModeCallback: ActionMode.Callback
    ) = ReleaseAdapter.create(
        recyclerView = binding.list,
        useLite = true,
        onSelectionChange = { size ->
            if (size == 0) {
                _listener.onFragmentRequestActionMode(ACTION_MODE_NAME)
            } else {
                _listener.onFragmentRequestActionMode(
                    ACTION_MODE_NAME,
                    getString(R.string.title_selected, size),
                    actionModeCallback
                )
            }
        },
        onReleaseClick = { release ->
            openEdit(binding.root, release)
        },
        onReleaseTogglePurchase = { release ->
            _viewModel.updatePurchased(
                release.release.id,
                !release.release.purchased
            )
        },
        onReleaseToggleOrder = { release ->
            _viewModel.updateOrdered(
                release.release.id,
                !release.release.ordered
            )
        }
    )

    private fun createComicsWithReleasesObserver(): Observer<ComicsWithReleases> {
        val context = requireContext()
        val txtInitial = binding.comics.txtComicsInitial
        val txtName = binding.comics.txtComicsName
        val txtPublisher = binding.comics.txtComicsPublisher
        val txtAuthors = binding.comics.txtComicsAuthors
        val txtNotes = binding.comics.txtComicsNotes
        val txtLast = binding.comics.txtComicsReleaseLast
        val txtNext = binding.comics.txtComicsReleaseNext
        val txtMissing = binding.comics.txtComicsReleaseMissing
        val txtComicsSourceId = binding.txtComicsSourceId

        return Observer { comics ->
            _comics = comics
            txtName.text = comics.comics.name
            txtPublisher.text = comics.comics.publisher
            txtAuthors.text = comics.comics.authors
            txtNotes.text = comics.comics.notes
            if (comics.comics.image != null) {
                txtInitial.text = ""
                Glide.with(this)
                    .load(Uri.parse(comics.comics.image))
                    .apply(ImageHelperKt.getInstance(context).circleOptions)
                    .into(DrawableTextViewTarget(txtInitial))
            } else {
                txtInitial.text = comics.comics.initial
            }
            val lastRelease = comics.lastPurchasedRelease
            txtLast.text =
                if (lastRelease == null) context.getString(R.string.release_last_none) else context.getString(
                    R.string.release_last,
                    lastRelease.number
                )
            val nextRelease = comics.nextToPurchaseRelease
            if (nextRelease != null) {
                if (nextRelease.date != null) {
                    // TODO: non mi piace, dovrei mostrare la data solo se futura e nel formato ddd dd MMM
                    txtNext.text = context.getString(
                        R.string.release_next_dated, nextRelease.number,
                        nextRelease.date.toHumanReadable(context)
                    )
                } else {
                    txtNext.text = context.getString(
                        R.string.release_next,
                        nextRelease.number
                    )
                }
            } else {
                txtNext.text = context.getString(R.string.release_next_none)
            }
            val missingCount = comics.notPurchasedReleaseCount
            txtMissing.text = context.getString(R.string.release_missing, missingCount)

            binding.comics.imgSourced.visibility = if (comics.comics.isSourced) View.VISIBLE else View.GONE

            if (BuildConfig.DEBUG) {
                txtComicsSourceId.visibility = View.VISIBLE
                txtComicsSourceId.text = comics.comics.sourceId ?: "unsourced"
            }
        }
    }

    private fun loadNewReleases() {
        _listener.resetUndo()
        binding.swipeRefresh.isRefreshing = true
        _viewModel.loadNewReleases(_comics)
    }

    private fun onNewReleasesLoaded(count: Int, tag: String) {
        LogHelper.d { "New releases: $count with tag '$tag'" }
        binding.swipeRefresh.isRefreshing = false
        if (count > 0) {
            Toast.makeText(
                requireContext(),
                resources.getQuantityString(
                    R.plurals.auto_update_available_message,
                    count,
                    count
                ),
                Toast.LENGTH_SHORT
            ).show()
        } else {
            Toast.makeText(
                requireContext(),
                R.string.auto_update_zero,
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    private fun onNewReleasesError() {
        binding.swipeRefresh.isRefreshing = false
        Toast.makeText(
            requireContext(),
            R.string.refresh_release_error,
            Toast.LENGTH_SHORT
        ).show()
    }

    private fun openEdit(view: View, release: ComicsRelease) {
        val directions: NavDirections = ComicsDetailFragmentDirections
            .actionDestComicsDetailFragmentToReleaseEditFragment(release.comics.id)
            .setReleaseId(release.release.id)
        findNavController(view).navigate(directions)
    }

    private fun shareReleases(releases: List<ComicsRelease>) {
        requireActivity().share(releases)
    }

    private fun onMarkedAsRemoved(count: Int, tag: String) {
        _listener.handleUndo(
            resources.getQuantityString(R.plurals.release_deleted, count, count),
            tag
        )
    }
}